### PR TITLE
feature: Add Framework Version support for PyTorch compilation (Neo)

### PR DESCRIPTION
**Pull request by prateekch91**
*Description of changes:*
Add Framework Version support for PyTorch compilation (Neo), also added image URI support for 1.5 and 1.6. Updated 1.4.0 to 1.4 in image URI for consistency.

```
Example:
model.compile(
        target_instance_family="ml_c4",
        input_shape={"data": [1, 3, 1024, 1024]},
        output_path="s3://output",
        role="role",
        framework="pytorch",
        framework_version="1.6",
        job_name="compile-model",
)
```
#### Testing
- Added test cases to test for different scenarios.
- Tested Sagemaker compile for PyTorch versions - 1.4/1.5/1.6 with new Boto changes.

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)

#### Tests

- [x ] I have added tests that prove my fix is effective or that my feature works (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/model.py [hunk 1]
```diff
@@ -407,6 +408,9 @@ def _compilation_job_config(
             else input_shape,
             "Framework": framework.upper(),
         }
+        if framework.upper() == "PYTORCH" and framework_version is not None:
```
Comment (lekk32) [inline src/sagemaker/model.py:411]: Please add a check for target_device is a cloud ioc target. Framework version is not supported for edge devices and ml_inf1.

Comment (prateekch91) [reply src/sagemaker/model.py:411]: Addressed. Thanks!!

File: src/sagemaker/model.py [hunk 2]
```diff
@@ -407,6 +408,19 @@ def _compilation_job_config(
...
+            if target_instance_type is not None:
+                if target_instance_type.startswith("ml_"):
+                    if "ml_inf" not in target_instance_type:
+                        input_model_config["FrameworkVersion"] = utils.get_short_version(
+                            framework_version
+                        )
+                    else:
+                        raise ValueError(
```
Comment (kkoppolu1) [inline src/sagemaker/model.py:419]: This seems to exclude deployment to ml_inf as well. Is that accurate?

Comment (prateekch91) [reply src/sagemaker/model.py:419]: Yes, when the model will not be compiled against this target, then there will be no deploy for this.

**Review by kkoppolu1 — APPROVED** [review]

**Review by lekk32 — APPROVED** [review]
Looks good to me. Thanks.

File: src/sagemaker/model.py [hunk 3]
```diff
@@ -407,6 +408,14 @@ def _compilation_job_config(
             else input_shape,
             "Framework": framework.upper(),
         }
+        if framework.upper() == "PYTORCH" and framework_version is not None:
+            if target_instance_type is not None:
+                if target_instance_type.startswith("ml_"):
+                    if "ml_inf" not in target_instance_type:
+                        input_model_config["FrameworkVersion"] = utils.get_short_version(
+                            framework_version
+                        )
```
Comment (metrizable) [inline src/sagemaker/model.py:417]: [Flat is better than nested](https://www.python.org/dev/peps/pep-0020/):

```suggestion
        if (
            framework.lower() == "pytorch" and
            framework_version is not None and
            target_instance_type is not None and
            target_instance_type.startswith("ml_") and
            "ml_inf" not in target_instance_type
        ):
            input_model_config["FrameworkVersion"] = utils.get_short_version(framework_version)
```

If you imported `re`, you could also provide a regex for the last two. Something like:

```
re.match("(?=^ml_)(?!ml_inf)", target_instance_type) is not None
```

File: src/sagemaker/model.py [hunk 4]
```diff
@@ -572,7 +581,8 @@ def compile(
             framework (str): The framework that is used to train the original
                 model. Allowed values: 'mxnet', 'tensorflow', 'keras', 'pytorch',
                 'onnx', 'xgboost'
-            framework_version (str):
+            framework_version (str):The version of framework, for example:
+                '1.5' for PyTorch
```
Comment (metrizable) [inline src/sagemaker/model.py:585]: space after colon. also, use e.g.?
```suggestion
            framework_version (str): The version of framework (e.g. '1.5' for PyTorch)
```

**Review by metrizable — CHANGES_REQUESTED** [review]
Thanks for the contribution! A few questions:

* Should you add an integration test?
* Should you unit test more combinations of inputs for the `_compilation_job_config` method to ensure proper behavior (specifically when `input_model_config["FrameworkVersion"]` is set)?

File: src/sagemaker/model.py (hunk 3 above)
Comment (prateekch91) [reply src/sagemaker/model.py]: Thanks for the review!! I have updated the changes accordingly.

File: src/sagemaker/model.py (hunk 4 above)
Comment (prateekch91) [reply src/sagemaker/model.py]: Thanks!! Addressed.

File: src/sagemaker/model.py [hunk 5]
```diff
@@ -36,6 +37,8 @@
     ["mxnet", "tensorflow", "keras", "pytorch", "onnx", "xgboost", "tflite"]
 )
 
+NEO_PYTORCH_DEFAULT_VERSION = "1.6"
```
Comment (metrizable) [inline src/sagemaker/model.py:40]: a default framework version was removed from the various frameworks in sdk v2 (july 2020) to promote explicitness and production stability

**Review by metrizable — CHANGES_REQUESTED** [review]

File: src/sagemaker/model.py [hunk 6]
```diff
@@ -407,6 +409,14 @@ def _compilation_job_config(
             else input_shape,
             "Framework": framework.upper(),
         }
+
+        if (
+            framework.lower() == "pytorch"
+            and re.match("(?=^ml_)(?!ml_inf)", target_instance_type) is not None
+        ):
+            if framework_version is not None:
+                input_model_config["FrameworkVersion"] = utils.get_short_version(framework_version)
```
Comment (metrizable) [inline src/sagemaker/model.py:418]:
```suggestion
        if (
            framework.lower() == "pytorch"
            and re.match("(?=^ml_)(?!ml_inf)", target_instance_type) is not None
            and framework_version is not None
        ):
            input_model_config["FrameworkVersion"] = utils.get_short_version(framework_version)
```

**Review by metrizable — CHANGES_REQUESTED** [review]

File: src/sagemaker/model.py (hunk 6 above)
Comment (prateekch91) [reply src/sagemaker/model.py:418]: Thanks for reviewing. I will make the suggested changes.

**Review by metrizable — APPROVED** [review]

**Pull request merged**